### PR TITLE
implemented ProductDaoJdbcImpl;

### PR DESCRIPTION
**Pull request by matviichuk92**

### CODE REVIEW

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            PreparedStatement statement =
+                    connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                return Optional.of(getProductFromResultSet(resultSet));
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find product by id!", e);
```
Comment (AnaLanda) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
            throw new DataProcessingException("Can't find product by id!" + id, e);
```
Maybe specifying the id that caused a problem in the exception message could help you debug later)

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            String query = "SELECT * FROM product_table;";
+            PreparedStatement statement = connection.prepareStatement(query);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                Product product = getProductFromResultSet(resultSet);
+                products.add(product);
+            }
+            return products;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get all product!", e);
```
Comment (AnaLanda) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get all products!", e);
```

**Review by AnaLanda — COMMENTED** [review]
Looks great!

File: src/main/java/online/shop/Application.java [hunk 3]
```diff
@@ -1,32 +1,18 @@
...
-        User roman = new User("roma","roma","123");
-        User mila = new User("mila","mila","123");
-        userService.create(roman);
-        userService.create(mila);
-        ShoppingCart shoppingCart = new ShoppingCart(roman.getId());
-        ShoppingCart shoppingCart1 = new ShoppingCart(mila.getId());
-        shoppingCartService.create(shoppingCart1);
-        shoppingCartService.create(shoppingCart);
-
+        ProductDaoJdbsImpl productDaoJdbs = new ProductDaoJdbsImpl();
```
Comment (boroda4436) [inline src/main/java/online/shop/Application.java]:
```suggestion
        ProductDao productDao = new ProductDaoJdbcImpl();
```
1. Use Abstraction
2. JDBC, not JDBS. Here `C` means connectivity

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.List;
+import java.util.Optional;
+import online.shop.dao.ProductDao;
+import online.shop.exception.DataProcessingException;
+import online.shop.lib.Dao;
+import online.shop.model.Product;
+import online.shop.util.ConnectionUtil;
+
+@Dao
+public class ProductDaoJdbsImpl implements ProductDao {
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
public class ProductDaoJdbcImpl implements ProductDao {
```

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            return product;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add item: " + product, e);
+        }
+    }
+
+    @Override
+    public Optional<Product> getById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM product_table WHERE product_id = ? AND deleted=false;";
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
            String query = "SELECT * FROM product_table WHERE product_id = ? AND deleted = false;";
```

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            PreparedStatement statement =
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
            PreparedStatement statement = connection.prepareStatement(query);
```
Can we inline declaration and initialization here?

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find product by id: " + id, e);
+        }
+    }
+
+    @Override
+    public List<Product> getAll() {
+        List<Product> products = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "SELECT * FROM product_table;";
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
            String query = "SELECT * FROM products;";
```
Do not append `_table`. We already know that this is a table.

Comment (boroda4436) [reply src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]: in all places

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            throw new DataProcessingException("Can't get all products!", e);
+        }
+    }
+
+    @Override
+    public Product update(Product product) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE product_table "
+                    + "SET name = ?, price = ? "
+                    + "WHERE product_id = ?;";
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
                    + "WHERE product_id = ? and deleted = false;";
```

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update product by id: " + product.getId(), e);
+        }
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "UPDATE product_table SET deleted=true "
+                    + "WHERE product_id = ? AND deleted=false;";
```
Comment (boroda4436) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]:
```suggestion
                    + "WHERE product_id = ? AND deleted = false;";
```

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/online/shop/Application.java (hunk 3 above)
Comment (matviichuk92) [reply src/main/java/online/shop/Application.java]: greatest shame(((

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java (hunk 7 above)
Comment (matviichuk92) [reply src/main/java/online/shop/dao/jdbc/ProductDaoJdbsImpl.java]: Product_table is the name of my table with products.

File: src/main/java/online/shop/dao/jdbc/ProductDaoJdbсImpl.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+import online.shop.lib.Dao;
+import online.shop.model.Product;
+import online.shop.util.ConnectionUtil;
+
+@Dao
+public class ProductDaoJdbсImpl implements ProductDao {
+    @Override
+    public Product create(Product product) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            String query = "INSERT INTO product_table (name, price) VALUES (?, ?);";
```
Comment (RomanDubovskyi) [inline src/main/java/online/shop/dao/jdbc/ProductDaoJdbсImpl.java]: this naming is redundant, it is obvious that statements are related to the tables, fix according to comment from Bohdan

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**